ocs_ci/utility/azure_utils.py

```diff
@@ -71,23 +93,80 @@ def get_node_by_attached_volume(self, volume):
 
         for vm in vm_list:
             for disk in vm.storage_profile.data_disks:
-                if disk.name == volume:
+                if disk.name == volume.name:
                     return vm
 
-    def detach_volume(self, volume, vm, timeout=120):
-        pass
+    def detach_volume(self, volume, node, timeout=120):
+        """
+        Detach volume if attached
 
+        Args:
+            volume (disk): disk object required to delete a volume
+            node (OCS): The OCS object representing the node
+            timeout (int): Timeout in seconds for API calls
 
-def get_data_volumes(deviceset_pvs):
-    """
-    Get the instance data volume names
+        """
+        vm = self.get_vm_instance(node.name)
+        data_disks = vm.storage_profile.data_disks
+        data_disks[:] = [disk for disk in data_disks if disk.name != volume.name]
+        logger.info(
+            "Detaching volume: %s Instance: %s", volume.name, vm.name
+        )
+        self.compute_client.virtual_machines.create_or_update(
+            self._resourcegroup,
+            vm.name,
+            vm)
+        for x in range(timeout):
+            disk_state = self.get_disk_state(volume.name)
+            logger.debug(
+                "Volume id: %s has status: %s", volume.name,
+                disk_state
+            )
+            if disk_state == 'Unattached':
+                break
+            if x == timeout - 1:
+                raise AZURETimeoutException(
+                    f"Reached timeout {timeout}s for volume to detach for "
+                    f"volume ID: {volume.name}, Volume state: "
+                    f"{disk_state}"
+                )
+            time.sleep(1)
+
+    def restart_az_vm_instance(self, vm_name):
+        """
+        Restart an Azure vm instance
 
-    Args:
-        deviceset_pvs (list): PVC objects of the deviceset PVs
+        Args:
+            vm_name: Name of azure vm instance
 
-    Returns:
-        list: Azure Vm Volume names
+        """
+        result = self.compute_client.virtual_machines.restart(
+            self._resourcegroup, vm_name)
+        result.wait()
 
-    """
-    volume_name = [pv.get().get('spec').get('azureDisk').get('diskName') for pv in deviceset_pvs]
-    return volume_name
+    def get_data_volumes(self, deviceset_pvs):
+        """
+        Get the instance data disk objects
+
+        Args:
+            deviceset_pvs (list): PVC objects of the deviceset PVs
+
+        Returns:
+            list: Azure Vm disk objects
+
+        """
+        volume_names = [pv.get().get('spec').get('azureDisk').get('diskName') for pv in deviceset_pvs]
+        return [self.compute_client.disks.get(self._resourcegroup, volume_name) for volume_name in volume_names]
```

Comment:
What is data type of `self.compute_client.disks` object?

Could you replace `.get('foo')` calls with `['foo']` when get is called on a dictionary? In case of a missing entry, it will raise an exception anyway, but exception caused by calling get() on a none object is less useful compared to missing key exception.

Reply:
**type of self.compute_client.disks :** `<azure.mgmt.compute.v2019_11_01.operations._disks_operations.DisksOperations object at 0x7f101b058c10>`

- Used index for accessing the dictionary instead of `.get('foo')`